### ISSUE TITLE
Downloading HydraX Streams

### ISSUE THREAD

**Issue by eslamsidereel** [thorio/KGrabber#35]
This script is amazing and after adding the idm export feature it became even more amazing. but since kissanime changed from rapidvideo to hydrax, i cant download any anime. I searched a lot but couldn't find a way to generate links from hydrax stream. It seems to be impossible so i had to use nova server and solve all the captcha to get the download links. If anybody knows a way to generate links from hydrax stream that would be a great help.

**Comment by thorio** [comment]
I tried (and failed) to do this a while ago. You can find my attempts on the [hydrax branch](https://github.com/thorio/KGrabber/tree/hydrax).
I got to the point of getting all the data they use to assemble the stream, but i wasn't able to assemble it or even play their streams outside of their website.
Perhaps someone more knowledgeable about HLS can crack this, in which case I'd be happy to write my findings down for you in detail.

**Comment by thorio** [comment]
Hydrax seems to have been removed. Re-open when it is spotted again.

**Comment by Soldat56** [comment]
Hey it is back...

**Comment by thorio** [comment]
Indeed it is.
I've re-added it to the script, although the links don't even work outside of an iframe anymore.

**Comment by Shubhodeep84** [comment]
Hydrax got The Alliance for creativity (ACE) protection which was created on 2017 they are anti-piracy Organization. They was even shutdown Openload and streaming at 31 October 2019. I guess Hydrax thought better form alliance get protected then shutdown as well.

**Comment by KartikSoneji** [comment]
@thorio 
I spent some time messing around with HydraX, and so far this is what I have been able to reverse engineer:

If the embed URL is `https://playhydrax.com/?v=VMG9wjJtM`

To download the video, first you need to make a POST request to `https://ping.idocdn.com/`.
```sh
curl "https://ping.idocdn.com/" -d "slug=VMG9wjJtM"
```

Response:
```json
{
    "status": true,
    "url": "m0weXd1eG1pdS54eXo=a",
    "sources": ["sd", "hd"],
    "isCdn": false,
    "isHls": false,
    "ads": "https://unreshiramor.com/4/3239564/"
}
```

Then, you can start downloading the video.
```sh
curl "https://www.4vs5osstge.buzz/" -H "Referer: https://playhydrax.com/?v=VMG9wjJtM" -o "video.mp4"
```

## Notes
- All the videos I found were mp4, so I don't know if this will work for HLS files.
  I can try it out if I get an HLS link.
- I have no idea how the quality selection works.
  Right now, this will download files in the highest quality available.

I haven't use IDM a lot, so I don't know if this method can be used with it.
PowerShell does have an in-built [`Invoke-WebRequest`](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/invoke-webrequest?view=powershell-7) command, and MacOS and Linux ship with `curl` so we can generate download scripts.

**Comment by JambDev** [comment]
It appears that the "www." in the URL actually determines it's quality. 
Ex. https://www.4vs5osstge.buzz/ would download 720p and https://4vs5osstge.buzz/ would download 360p

I still have no idea how it deciphers the link from "m0weXd1eG1pdS54eXo=a"

**Comment by thorio** [comment]
`am0weXd1eG1pdS54eXo=` decodes to `jm0ywuxmiu.xyz`, which is different from `4vs5osstge.buzz` but still works. Maybe the url doesn't actually matter? When you open the dev console on the kissanime page it tells you it's a google drive proxy, so the servers might all be identical.

@KartikSoneji I don't think IDM supports Headers outside of the browser extensions.

**Comment by KartikSoneji** [comment]
@JambDev 
> It appears that the "www." in the URL actually determines it's quality.

Hmm, quite clever.

> EDIT: I actually did end up figuring out how to "decode" the URL provided by ping.idocdn.com
> All you have to do is move the last character in the string to the beginning and then decode it in base64.

Great!
How did you figure it out? I would never have thought that a company would use something like this.
Looks like HydraX is going for security through obscurity.

Still haven't been able to test HSL though.

@thorio 
Interestingly, you don't have to use the url returned from `https://ping.idocdn.com`.
For example, I used `4vs5osstge.buzz` for all requests, and it worked for ~140 videos.
If you resolve the host with `nslookup`, you will find that they point to different ip addresses.
Maybe the multiple urls are a form of load-balancing?
But it is good that we can decode them now.

> @KartikSoneji I don't think IDM supports Headers outside of the browser extensions.

I was afraid of that.
There has been some discussion about [allowing custom headers in VLC](https://forum.videolan.org/viewtopic.php?f=29&t=145323) for remote playback, but it does not seem to be implemented currently.

How about generating a PowerShell/cmd/shell script to download the videos?
I still need to work out some details (like `hd`/`sd` selection and fixing the base64) but this is what I have so far:
```shell
# downloadHydraXVideo <slug> <output file>
downloadHydraXVideo(){	
	url64=$(curl "https://ping.idocdn.com/" -d "slug=$1" | grep -Po '(?<="url":")[a-zA-Z0-9=+-]*?(?=",)');
	
	# TODO: Fix the base64 string by moving the last character to the front.
	curl "$(echo $url64 | base64 -d)" -# -C - -H "Referer: https://playhydrax.com/?v=$1" -o "$2";
}

downloadHydraXVideo "VMG9wjJtM" "Video 1.mp4"
downloadHydraXVideo "TiDe3FPHK" "Video 2.mp4"
#...
```

**Comment by thorio** [comment]
@KartikSoneji I believe aria2c supports custom headers, as well as multithreaded downloads. I looked into it a while ago, I don't remember why I didn't pursue it further. We should be able to export a file with all the urls and headers required.

**Comment by KartikSoneji** [comment]
@thorio Agree, but that would be an external dependency.
`curl` is installed on almost all systems, and `Invoke-WebRequest` ships with all modern versions of Windows.
But it can be good as an additonal option though.

**Comment by thorio** [comment]
There are already aria2c and IDM exporters in the script. I don't believe curl can be multithreaded on the same download?
I'll go ahead and add a step to ping `idocdn` and get the url, as well as a new aria2c exporter for this.
Feel free to open a PR for the curl version.

**Comment by KartikSoneji** [comment]
> I don't believe curl can be multithreaded on the same download?

Unfortunately, no :(

> I'll go ahead and add a step to ping idocdn and get the url

If there is a large delay (> 5-6 hours) between pinging idocdn and downloading the video then you *might* get an `Access Denied` error.
The only way that worked reliably is to ping idocnd right before trying to download the video.

I don't know if that is possible in `aria2c` without some shell scripting.

> Feel free to open a PR for the curl version.

OK, will do that.

**Comment by thorio** [comment]
I pushed the action to [feature/hydrax-download](https://github.com/thorio/KGrabber/tree/feature/hydrax-download).
You can copy one of the existing exporters and edit it accordingly.

**Comment by KartikSoneji** [comment]
@thorio Thanks, will take a look.

**Comment by shinji257** [comment]
I wrote up a quick bash script that you can provide the stub to and it will retrieve the video.  It assumes 720p for the video resolution.  I don't know what would happen if that wasn't available and wasn't tested.  If you were grabbing a series that only had a 360p version you can remove the www. in the url that it accesses in order to get that one.  I might implement a check that handles it.  Not sure.  For now just remove the www. in the url if you know you can't get a 720p version.

--insecure is needed on the second curl statement as it seems that at least one of the domains to download the video from doesn't have a valid SSL certificate so curl aborted on it.

```
#!/usr/bin/env bash

# Video stub
STUB=$1

# Output filename.  If none set in command then default to stub.mp4
OUT=$2
if [ -z "$OUT" ]; then OUT="$STUB.mp4"; fi

# Grab JSON data from server
JSON=`curl -s "https://ping.idocdn.com/" -d "slug=$STUB"`

# Cut away and eventually get domain to download video from. -- Load balancing?
# A bit messy here.  First cut away excess data and remove enclosing quotes.
# After use a bit of possible bashism to rotate string so that the last character is in front of first character.
# Finally store the URL for the retrieval.
URLB64=`echo $JSON | cut -d, -f2 | cut -d: -f2`
URLB64=`sed -e 's/^"//' -e 's/"$//' <<<"$URLB64"`
LURL=${#URLB64}
URLB64=${URLB64:$LURL-1:1}${URLB64:0:$LURL-1}
URL=`echo $URLB64 | base64 -d`

# Remove www. in the url below to get 360p version instead of 720p.
# There is no option to give the script yet to toggle it directly.

curl --insecure "https://www.$URL/" --referer "https://playhydrax.com/?v=$STUB" -o "$OUT"
```

**Comment by jazamb** [comment]
I've noticed that the requests HydraX's player is making don't directly fetch the mp4 anymore. 
At the time of writing this, the current method (in the replies to this issue) still works. I am afraid that it might break though with this new web socket approach they are using

**Comment by shinji257** [comment]
This is no longer needed. KissAnime.ru shut down permanently a few minutes ago.

**Comment by jazamb** [comment]
welp

**Comment by thorio** [comment]
@JambDev When I originally tried this (before this issue was opened) they were using HLS, so they might be changing things around a lot.

@shinji257 There is still kissanime.nz (#39), which also has a hydrax option (Hserver). They also don't seem to have captchas?
I will look at adding support for it again when I get the chance, but any help is appreciated.

**Comment by shinji257** [comment]
@thorio Ahh. I was unaware that it also used hydrax.

**Comment by thorio** [comment]
@shinji257 Yes, and if you look closely you can still find the familiar ping to `idocdn` in the network tab.
So it's just a matter of figuring out how they get the link to the iframe.

**Comment by KartikSoneji** [comment]
@thorio @shinji257 

`kissanime.nz` is down as well, so maybe the outage is temporary?

![image](https://user-images.githubusercontent.com/25801283/90313154-da746080-df27-11ea-99ed-853907388ede.png)

**Comment by shinji257** [comment]
KissAnime.nz is still working here.

**Comment by KartikSoneji** [comment]
Yay, `kissanime.nz` works for me now!

**Comment by KartikSoneji** [comment]
I looked at `https://kisscartoon.nz/` and `https://kimcartoon.si/` (#46)

For `https://kisscartoon.nz/Cartoon/Angry-Birds.88861/Movie-HD?id=71028`

```sh
# server = hserver | xserver | ptserver | oserver
curl "https://kisscartoon.nz/ajax/anime/load_episodes_v2?s=<server>" -d "episode_id=71028"
```

Using `hserver`:
```sh
curl "https://kisscartoon.nz/ajax/anime/load_episodes_v2?s=hserver" -d "episode_id=71028"
```

```json
{
    "status": true,
    "value": "<iframe src=\"https:\/\/embed.streamx.me?k=756ecb2dfb240248235934ab3ac1b943&li=71028&tham=1597515866&lt=hydrax&check_hot=1&qlt=720p&spq=p&prv=bWVkaWEvdGh1bWIvMTcwNDA4XzA1NDYvQW5ncnktQmlyZHMuODgxNTktTW92aWUtSEQuanBnO21lZGlhL3RodW1iLzE3MDQwOF8wNTQ2L0FuZ3J5LUJpcmRzLjg4MTU5LU1vdmllLUhEX21vYmlsZS5qcGc%3D&key=57443b94a0716a10f23334135e9979e1&ua=5f2bf4be52e6a7bdd37bca5ebabb7b2c\" width=\"100%\" height=\"100%\" style=\"border:none;overflow:hidden\" scrolling=\"no\" frameborder=\"0\" allowTransparency=\"true\" allowFullScreen=\"true\" allow=\"accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture\"><\/iframe>",
    "embed": true,
    "html5": false,
    "type": "",
    "sv": "hydrax",
    "download_get": ""
}
```

Using the url from the iframe's `src`.
(`referer` is needed)
```sh
curl \
    --referer "https://kisscartoon.nz/" \
    "https://embed.streamx.me?k=756ecb2dfb240248235934ab3ac1b943&li=71028&tham=1597515866&lt=hydrax&check_hot=1&qlt=720p&spq=p&prv=bWVkaWEvdGh1bWIvMTcwNDA4XzA1NDYvQW5ncnktQmlyZHMuODgxNTktTW92aWUtSEQuanBnO21lZGlhL3RodW1iLzE3MDQwOF8wNTQ2L0FuZ3J5LUJpcmRzLjg4MTU5LU1vdmllLUhEX21vYmlsZS5qcGc%3D&key=57443b94a0716a10f23334135e9979e1&ua=5f2bf4be52e6a7bdd37bca5ebabb7b2c"
```

returns a lot of HTML, but the interesting bit is
```html
<script>
    new Player({
        "key": "e6d3ce91839c03915de9125be5ed1f47",
        "type": "slug",
        "value": "3aOgVSRom39",
        "width": "100%",
        "height": "100%",
        "cast": "{}",
        "autostart": true,
        tracks: [[]]
    });
</script>
```

Which finally gives us the `slug` as `3aOgVSRom39`.

The same works for `https://kimcartoon.si/`, simply by replacing `kisscartoon.nz` with `kimcartoon.si`.

Update: `https://kissanime.nz` also seems to use the same setup, but they have a LOT of scraping protection (captchas, csrf tokens).

**Comment by Thanhcaro** [comment]
Please help me download this type of link: https://playhydrax.com/?v=GDUc-uLQN
Sorry I don't know how to use script, I install on Chrome and open website but still can not load video.

**Comment by KartikSoneji** [comment]
Hi @Thanhcaro 
Please refer to the usage instructions: https://github.com/thorio/KGrabber#installation-and-usage

**Comment by Thanhcaro** [comment]
> Hi @Thanhcaro
> Please refer to the usage instructions: https://github.com/thorio/KGrabber#installation-and-usage

The script is currently compatible with:

kissanime.ru
kimcartoon.to
kissasian.sh
kisstvshow.to

my website is not in this list.

**Comment by KartikSoneji** [comment]
@Thanhcaro 

https://github.com/thorio/KGrabber#compatibility
> If you'd like to add a site, please open a new issue on GitHub.

Can you open an issue?

**Comment by Thanhcaro** [comment]
> @Thanhcaro
> 
> https://github.com/thorio/KGrabber#compatibility
> 
> > If you'd like to add a site, please open a new issue on GitHub.
> 
> Can you open an issue?

Oh Sorry. I will open new issue. Thanks for your help.

**Comment by Thanhcaro** [comment]
Sorry, i thought script can download all hydrax link from any website. I
already open new issue, pls do me a favour. Thank you.

Vào 00:46, T.3, 6 Th10, 2020 Kartik Soneji <notifications@github.com> đã
viết:

**Comment by mrwunderbar666** [comment]
I have been lurking this issue for a while now and was able to implement the hydrax downloader in Python.

But recently, something about the `https://ping.iamcdn.net/` has changed. It now returns a `502` error.

Using the question from #54 as an example:

```
curl "https://ping.iamcdn.net/" -d "slug=7fUvtRbiYk"
```

Returns:
```
<html>
<head><title>502 Bad Gateway</title></head>
<body>
<center><h1>502 Bad Gateway</h1></center>
<hr><center>nginx/1.19.6</center>
</body>
</html>
```

However, it appears that the following steps remained the same. I could download files via trial and error by using the old domains such as: `curl "https://x00vfu277x.buzz/" -H "Referer: https://playhydrax.com/?v=TZTopeD-V" -o "test01.mp4"`

I tried to dig in the [js files](https://iamcdn.net/players/playhydrax.dev.min.js) for clues, but couldn't find anything :/

**Comment by KartikSoneji** [comment]
@mrwunderbar666 Seems like the `ping.iamcdn.net` step is no longer necessary.

Running:
```sh
curl -L https://playhydrax.com/?v=TZTopeD-V
```

The interesting bit:
```html
<script type="text/javascript">
    //...
    if ("undefined" != typeof ADS)
        new ADS(atob("eyJzdGF0dXMiOnRydWUsIm1kNV9pZCI6MTAzNTc3MjUsInNvdXJjZXMiOlsic2QiLCJoZCJdLCJzb3VyY2VfYnJlYWtzIjpbXSwiZG9tYWluIjoiNDd5YW9odWhpbi5idXp6IiwidXNlX3AycCI6dHJ1ZSwidXNlX2Rvd24iOmZhbHNlfQ=="))

    var styleMinion = "width:100%;height:100% !important;background: url(https://iamcdn.net/img/block2.jpg) center center no-repeat;background-size:contain;position:absolute"
</script>
```

decoding the base64 data gives:
```json
{
    "status": true,
    "md5_id": 10357725,
    "sources": ["sd", "hd"],
    "source_breaks": [],
    "domain": "47yaohuhin.buzz",
    "use_p2p": true,
    "use_down": false
}
```

Shell script:
```sh
curl -sL "https://playhydrax.com/?v=TZTopeD-V" |\
    grep "ADS(" |\
    sed -r 's/.*"(.*)".*/\1/' |\
    base64 -di
```

**Comment by mrwunderbar666** [comment]
@KartikSoneji Thanks that did the trick!

That is now much easier to implement. 

I just noticed that some of the domains have issues with SSL. No matter how I connect to some of them, some throw a handshake error:

```
HTTPSConnectionPool(host='www.x0g2gwwwr7.xyz', port=443): Max retries exceeded with url: / (Caused by SSLError(SSLError(1, '[SSL: SSLV3_ALERT_HANDSHAKE_FAILURE] sslv3 alert handshake failure (_ssl.c:1123)')))
```

I wonder, how the browser resolves that, because in curl the option `--insecure` does not help for handshake errors

**Comment by Thanhcaro** [comment]
> @KartikSoneji @mrwunderbar666 Thanks that did the trick!
> 
> That is now much easier to implement.
> 
> I just noticed that some of the domains have issues with SSL. No matter how I connect to some of them, some throw a handshake error:
> 
> ```
> HTTPSConnectionPool(host='www.x0g2gwwwr7.xyz', port=443): Max retries exceeded with url: / (Caused by SSLError(SSLError(1, '[SSL: SSLV3_ALERT_HANDSHAKE_FAILURE] sslv3 alert handshake failure (_ssl.c:1123)')))
> ```
> 
> I wonder, how the browser resolves that, because in curl the option `--insecure` does not help for handshake errors

You are so talent. Can you tell me the way to get hydrax link to download?

**Comment by kukhanh29** [comment]
> > @KartikSoneji @mrwunderbar666 Thanks that did the trick!
> > That is now much easier to implement.
> > I just noticed that some of the domains have issues with SSL. No matter how I connect to some of them, some throw a handshake error:
> > ```
> > HTTPSConnectionPool(host='www.x0g2gwwwr7.xyz', port=443): Max retries exceeded with url: / (Caused by SSLError(SSLError(1, '[SSL: SSLV3_ALERT_HANDSHAKE_FAILURE] sslv3 alert handshake failure (_ssl.c:1123)')))
> > ```
> > 
> > 
> > I wonder, how the browser resolves that, because in curl the option `--insecure` does not help for handshake errors
> 
> You are so talent. Can you tell me the way to get hydrax link to download?

A fool does not read the comments above.

**Comment by kukhanh29** [comment]
wow. seems it was blocked by encryption js

**Comment by tienduynguyen318** [comment]
it seems that hydrax.net has become abyss.to. Can anyone still download the files with the curl? If you can, please share the result.

Thanks

**Comment by lavara123** [comment]
@KartikSoneji  Do you know about abyss.to? how to do

**Comment by KartikSoneji** [comment]
@lavara123 If you find a website that uses abyss.to, then I'll be happy to update the script.

**Comment by lavara123** [comment]
> @lavara123 If you find a website that uses abyss.to, then I'll be happy to update the script.

Thanks a lot, I'm looking forward to it

```sh
curl 'https://embed2.megaxfer.ru/embed2/01bded490f99430277078b8ad5ae9af8' \
  -H 'user-agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.114 Safari/537.36' \
  -H 'referer: https://www.2embed.ru/' \
  --compressed
```


curl output:
```html
    <iframe src="https://playhydrax.com/?v=2FKgbNJtB&sub=https://cc.1clickcdn.ru/8d/3b/8d3b7506346ba4ef945e5274488b094e/eng-6.vtt" width="100%" height="100%" scrolling="no" frameborder="0"
            allowFullScreen="true" webkitallowfullscreen="true"
            mozallowfullscreen="true"></iframe>
```

and next link :
https://geoip.redirect-ads.com/?v=2FKgbNJtB

**Comment by KartikSoneji** [comment]
Looks like they are using CloudFlare, this is not going to be as easy.

**Comment by lavara123** [comment]
> Looks like they are using CloudFlare, this is not going to be as easy.

```sh
wget -d --header="User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.77 Safari/537.36" https://playhydrax.com/?v=2FKgbNJtB
```

**Comment by crypto-0** [comment]
@KartikSoneji  @lavara123  did you ever figure out how to bypass the cloudflare I have been stuck for days in figuring out that section I don’t know what checks are being performed

**Comment by KartikSoneji** [comment]
Hi @crypto-0 
Before, I was able to bypass Cloudflare's captcha with these options, but I haven't tested it recently.
```sh
curl -L --user-agent "Mozilla/5.0" --cipher AES256-SHA256 --tls-max 1.2 "<url>"
```

In nodejs, you have to set the cipher explictly:
```js
const https = require("https");

async function fetch(url, options){
    return new Promise((resolve, reject) => {
        let req = https.request(url, options, res => {
            let data = [];
            res.on("data", chunk => data.push(chunk));
            res.on("end", e =>
                resolve({
                    headers: res.headers,
                    status: res.statusCode,
                    text: async () => Buffer.concat(data).toString()
                })
            );
        });
        
        req.on("error", reject);
        
        req.end();
    });
}

// use
let response = await fetch(url, {
    headers: { "User-Agent": "Mozilla/5.0" },
    ciphers: "AES256-SHA256"
});
```

Hope that helps.

**Comment by kukhanh29** [comment]
> Hi @crypto-0 Before, I was able to bypass Cloudflare's captcha with these options, but I haven't tested it recently.
> 
> ```shell
> curl -L --user-agent "Mozilla/5.0" --cipher AES256-SHA256 --tls-max 1.2 "<url>"
> ```
> 
> In nodejs, you have to set the cipher explictly:
> 
> ```js
> const https = require("https");
> 
> async function fetch(url, options){
>     return new Promise((resolve, reject) => {
>         let req = https.request(url, options, res => {
>             let data = [];
>             res.on("data", chunk => data.push(chunk));
>             res.on("end", e =>
>                 resolve({
>                     headers: res.headers,
>                     status: res.statusCode,
>                     text: async () => Buffer.concat(data).toString()
>                 })
>             );
>         });
>         
>         req.on("error", reject);
>         
>         req.end();
>     });
> }
> 
> // use
> let response = await fetch(url, {
>     headers: { "User-Agent": "Mozilla/5.0" },
>     ciphers: "AES256-SHA256"
> });
> ```
> 
> Hope that helps.

Thank for share.
how to download stream ?
Can you do it ?

**Comment by Nicryc** [comment]
Any chance someone found a working solution to this issue?

**Comment by metzger100** [comment]
@KartikSoneji Sorry for Pinging you, but I am searching for method to download content from abyss.to. To give an example Movie: https://topstreamfilm.org/player.html?v=jxx2_qNnl. I searched for a long time now in the Internet for a working method but I didn't find any. 
The Adress of the Movie is freeplayervideo.com/?v=jxx2_qNnl but I have no Idea how to get a working download from it. I am relativley new to Web Reverse Engineering and I know the methods to download Files directly with rightclickenabler or Monitoring the Networktraffic to get a .m3u8-File to download the Movie with ffmpeg. Would be awesome if you could help me understanding what abyss.to does and how I can get the Download for movies hosted by abyss.to. I checked the Docs of JW-Player but that didn't help me a lot. Greetings Metzger.

**Comment by PatrickL546** [comment]
I made a repo since it's too long [How-to-download-hydrax-abyss.to](https://github.com/PatrickL546/How-to-download-hydrax-abyss.to)

**Comment by lavara123** [comment]
> I found a solution on this specific site that uses Hydrax/Abyss.to Hope this helps!
> 
> ```

Is it possible to do scraping?

**Comment by lavara123** [comment]
> @lavara123 Seems very possible to automate with little script. The waiting on globalcdn to appear might be sped up by expiring it faster or making it reconnect faster, somehow.

How to get here with just the embed url in scraping?

**Comment by PatrickL546** [comment]
@lavara123 Haven't looked into automating or scraping with it so can't help ya

**Comment by rosshauptnerin** [comment]
~~The videocdn URL does not show up for me. Only wss://connect.idocdn.com:3010/ and wss://symphony-riding-higher-each.trycloudflare.com/~~

~~What am I doing wrong? I am using it for the same site as you btw (gojo2)~~

~~EDIT NVM. I got the URL to show now, but if I try to run the code I get~~
  
~~File "C:\Users\XXXX\PycharmProjects\XXXX\.venv\Lib\site-packages\requests\sessions.py", line 794, in get_adapter
    raise InvalidSchema(f"No connection adapters were found for {url!r}")
requests.exceptions.InvalidSchema: No connection adapters were found for 'wss://gpcsyjfe71.globalcdn93.one/e11443fc4ace6821b3'~~

EDIT2: NVM got it to work now. I had to replace wss with https. But still, is there a quicker and easier version to get the videocdn URL (wss://gpcsyjfe71.globalcdn93.one) other than disconnecting and reconnecting from the internet so I can see and copy it from the websocket tab?

**Comment by PatrickL546** [comment]
New method only needs one link. More info https://github.com/PatrickL546/How-to-download-hydrax-abyss.to

` pip install requests --upgrade`

![image](https://github.com/PatrickL546/How-to-download-hydrax-abyss.to/assets/75874561/a63f88d6-4e6f-4237-bfe0-1692d1a3d315)

```Python
from base64 import b64decode
from requests import get
from json import loads
from re import search

print('cdn_ID. ie. "?v=VswFqVUmq" without "?v=". Note: ID should be 9 characters')
cdn_ID = input("Enter cdn_ID: ")

# leave empty for 360p
# add "www" for 720p
# add "whw" for 1080p
# if 1080p is not available, it will use the next highest quality
q_prefix = "whw"

domain, vid_id, sources = [loads(b64decode(search(r'PLAYER\(atob\("(.*?)"', get(f"https://abysscdn.com/?v={cdn_ID}").text).group(1)))[i] for i in ["domain","id","sources"]]

print(f"""
360p  =  " "  = "sd", "mHd"
720p  = "www" = "hd"
1080p = "whw" = "fullHd"

Available sources {sources}
Downloading "{q_prefix}" or next highest source available
Please wait...
""")

response = get(f"https://{domain}/{q_prefix}{vid_id}", headers={"Referer": f"https://abysscdn.com/?v={cdn_ID}"}, stream=True)
with open(f"{cdn_ID}.mp4", "wb") as f:
    for chunk in response.iter_content(chunk_size=8192):
        f.write(chunk)
```

**Comment by kukhanh29** [comment]
How do I get it to show download information?

**Comment by rosshauptnerin** [comment]
Thank you so much for you work and all the help so far! I have another quick question, what would I have to add to the code so I can queue up multiple videos to download? I'd love to be able to just plug in multiple video IDs and then have it download them while I can do other things rather than having to start one download after the other. Or is that just not possible?

**Comment by PatrickL546** [comment]
@rosshauptnerin . Quick and dirty script update

Check if download file already exists
Multiple ID input, removes duplicate ID's from input
Print errors, and some error handling

```Python
from base64 import b64decode
from os.path import isfile
from requests import get
from json import loads
from re import search

print(
    'cdn_ID. ie. "?v=VswFqVUmq" without "?v=". Note: ID should be 9 characters. Separate ID with comma ","'
)
cdn_ID_list = set([i.strip() for i in input("Enter cdn_ID: ").split(",")])

# leave empty for 360p
# add "www" for 720p
# add "whw" for 1080p
# if 1080p is not available, it will use the next highest quality
q_prefix = "whw"

error = []
for cdn_ID in cdn_ID_list:
    try:
        if isfile(f"./{cdn_ID}.mp4"):
            print(f"{cdn_ID}.mp4 already exists")
        else:
            domain, vid_id, sources = [
                loads(
                    b64decode(
                        search(
                            r'PLAYER\(atob\("(.*?)"',
                            get(f"https://abysscdn.com/?v={cdn_ID}").text,
                        ).group(1)
                    )
                )[i]
                for i in ["domain", "id", "sources"]
            ]

            print(f"""
360p  =  " "  = "sd", "mHd"
720p  = "www" = "hd"
1080p = "whw" = "fullHd"

Available sources {sources}
Downloading "{cdn_ID}.mp4" in quality "{q_prefix}" or next highest source available
Please wait...
""")

            response = get(
                f"https://{domain}/{q_prefix}{vid_id}",
                headers={"Referer": f"https://abysscdn.com/?v={cdn_ID}"},
                stream=True,
            )
            with open(f"{cdn_ID}.mp4", "wb") as f:
                for chunk in response.iter_content(chunk_size=8192):
                    f.write(chunk)
    except Exception:
        error.append(cdn_ID)

for i in error:
    print(f"Error downloading: {i}")
```

**Comment by rosshauptnerin** [comment]
> @rosshauptnerin . Quick and dirty script update
> 
> Check if download file already exists Multiple ID input, removes duplicate ID's from input Print errors, and some error handling
> 
> ```python
> from base64 import b64decode
> from os.path import isfile
> from requests import get
> from json import loads
> from re import search
> 
> print(
>     'cdn_ID. ie. "?v=VswFqVUmq" without "?v=". Note: ID should be 9 characters. Separate ID with comma ","'
> )
> cdn_ID_list = set([i.strip() for i in input("Enter cdn_ID: ").split(",")])
> 
> # leave empty for 360p
> # add "www" for 720p
> # add "whw" for 1080p
> # if 1080p is not available, it will use the next highest quality
> q_prefix = "whw"
> 
> error = []
> for cdn_ID in cdn_ID_list:
>     try:
>         if isfile(f"./{cdn_ID}.mp4"):
>             print(f"{cdn_ID}.mp4 already exists")
>         else:
>             domain, vid_id, sources = [
>                 loads(
>                     b64decode(
>                         search(
>                             r'PLAYER\(atob\("(.*?)"',
>                             get(f"https://abysscdn.com/?v={cdn_ID}").text,
>                         ).group(1)
>                     )
>                 )[i]
>                 for i in ["domain", "id", "sources"]
>             ]
> 
>             print(f"""
> 360p  =  " "  = "sd", "mHd"
> 720p  = "www" = "hd"
> 1080p = "whw" = "fullHd"
> 
> Available sources {sources}
> Downloading "{cdn_ID}.mp4" in quality "{q_prefix}" or next highest source available
> Please wait...
> """)
> 
>             response = get(
>                 f"https://{domain}/{q_prefix}{vid_id}",
>                 headers={"Referer": f"https://abysscdn.com/?v={cdn_ID}"},
>                 stream=True,
>             )
>             with open(f"{cdn_ID}.mp4", "wb") as f:
>                 for chunk in response.iter_content(chunk_size=8192):
>                     f.write(chunk)
>     except Exception:
>         error.append(cdn_ID)
> 
> for i in error:
>     print(f"Error downloading: {i}")
> ```

That's perfect thank you!

**Comment by PatrickL546** [comment]
I'm pretty happy with the current state of these
# [Hydrax-Abyss.to-DownloadHelper-Userscript](https://github.com/PatrickL546/Hydrax-Abyss.to-DownloadHelper-Userscript)

![image](https://github.com/thorio/KGrabber/assets/75874561/92e66288-3c36-48db-99f4-e0b798cbe803)

# [Hydrax-Abyss.to-DownloadHelper-Python](https://github.com/PatrickL546/Hydrax-Abyss.to-DownloadHelper-Python)

![image](https://github.com/thorio/KGrabber/assets/75874561/1288bb1f-6309-4a82-b6d5-ad1b9db15467)

# Just info [How-to-download-hydrax-abyss.to](https://github.com/PatrickL546/How-to-download-hydrax-abyss.to)
